apps/full-site-editing/full-site-editing-plugin/starter-page-templates/class-wp-rest-sideload-image-controller.php

```diff
@@ -0,0 +1,149 @@
+<?php
+/**
+ * WP_REST_Sideload_Image_Controller file.
+ *
+ * @package full-site-editing
+ */
+
+/**
+ * Class WP_REST_Sideload_Image_Controller.
+ */
+class WP_REST_Sideload_Image_Controller extends WP_REST_Attachments_Controller {
+
+	/**
+	 * WP_REST_Sideload_Image_Controller constructor.
+	 */
+	public function __construct() {
+		$this->namespace = 'fse/v1';
+		$this->rest_base = 'sideload/image';
+	}
+
+	/**
+	 * Register available routes.
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			[
+				[
+					'methods'             => WP_REST_Server::CREATABLE,
+					'callback'            => [ $this, 'create_item' ],
+					'permission_callback' => [ $this, 'create_item_permissions_check' ],
+					'show_in_index'       => false,
+					'args'                => [
+						'url'     => [
+							'description' => 'URL to the image to be side-loaded.',
+							'type'        => 'string',
+							'format'      => 'uri',
+							'required'    => true,
+						],
+						'post_id' => [
+							'description' => 'ID of the post to associate the image with',
+							'type'        => 'integer',
+							'default'     => 0,
+						],
+					],
+				],
+				'schema' => [ $this, 'get_item_schema' ],
+			]
+		);
+	}
+
+	/**
+	 * Creates a single attachment.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_Error|WP_REST_Response Response object on success, WP_Error object on failure.
+	 */
+	public function create_item( $request ) {
+		if ( in_array( get_post_type( $request->get_param( 'post_id' ) ), [ 'revision', 'attachment' ], true ) ) {
+			return new WP_Error( 'rest_invalid_param', __( 'Invalid parent type.' ), [ 'status' => 400 ] );
+		}
+
+		// Include image functions to get access to wp_read_image_metadata().
+		require_once ABSPATH . 'wp-admin/includes/file.php';
+		require_once ABSPATH . 'wp-admin/includes/image.php';
+		require_once ABSPATH . 'wp-admin/includes/media.php';
+
+		// The post ID on success, WP_Error on failure.
+		$id = media_sideload_image(
+			$request->get_param( 'url' ),
+			$request->get_param( 'post_id' ),
+			null,
```

Comment:
Maybe it worths adding the possibility to set up the image title, through of a `title` parameter for instanance:x

```php
		$id = media_sideload_image(
			$request->get_param( 'url' ),
			$request->get_param( 'post_id' ),
			$request->get_param( 'title' ),
			'id'
		);
```

Reply:
I don't think we currently have that information in our templates. Not sure how generic we want to make this endpoint?

Reply:
Yeah, in fact, if we want to make it more generic we should consider changing the namespace.
Anyway, there's anything wrong to allow the `title` attr? In the worse of cases, we just ignore it.